File: arrays/290. Word Pattern/wordPattern.js
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
var wordPattern = function(pattern, s) {
    let d = {};
    let words = s.split(" ");

    if (pattern.length !== words.length) return false;

    for (let i = 0; i < pattern.length; i++) {
        if (pattern[i] in d) {
            if (d[pattern[i]] !== words[i]) return false;
        } else {
            if (Object.values(d).includes(words[i])) return false;
            d[pattern[i]] = words[i];
        }
    }
    return true;
};

File: arrays/290. Word Pattern/wordPattern.py
class Solution(object):
    def wordPattern(self, pattern, s):
        """
        :type pattern: str
        :type s: str
        :rtype: bool
        """
        # dictionary to map letters in pattern to words in s
        d = {}
        # split the string into words
        words = s.split(" ")

        if len(pattern) != len(words):
            return False
        
        for i in range(len(pattern)):
            if pattern[i] in d:
                # if the letter in the dictionary is mapped to a different word
                if d[pattern[i]] != words[i]:
                    return False
            else:
                # if the word is already used for another letter
                if words[i] in d.values():                   
                    return False
                d[pattern[i]] = words[i]
        
        return True